nemo/collections/tts/modules/audio_codec_modules.py

```diff
@@ -2423,3 +2550,303 @@ def forward(self, audio, audio_len):
+            kernel_size=kernel_size,
+            activation=activation
+        )
+        self.pre_res_block = ResidualBlockV2(
+            channels=input_filters,
+            filters=input_filters,
+            kernel_size=kernel_size,
+            activation=activation
+        )
+        input_dim = input_filters
+        self.stft_blocks = nn.ModuleList([])
+        for resolution, filters in zip(resolutions[1:], resolution_filter_list[1:]):
+            stft_block = STFTResidualBlock(
+                resolution=resolution,
+                input_dim=input_dim,
+                down_sample_rate=2,
+                filters=filters,
+                kernel_size=kernel_size,
+                activation=activation,
+            )
+            self.stft_blocks.append(stft_block)
+            input_dim = filters
+
+        if down_sample_filter_list and not down_sample_rate_list:
+            down_sample_rate_list = len(down_sample_filter_list) * [2]
+
+        self.down_sample_blocks = nn.ModuleList([])
+        for (filters, down_sample_rate) in zip(down_sample_filter_list, down_sample_rate_list):
+            down_sample_block = DownSampleResidualBlock(
+                channels=input_dim,
+                filters=filters,
+                down_sample_rate=down_sample_rate,
+                kernel_size=kernel_size,
+                activation=activation
+            )
+            self.down_sample_blocks.append(down_sample_block)
+            input_dim = filters
+
+        self.post_conv = Conv1dNorm(
+            in_channels=input_dim,
+            out_channels=out_dim,
+            kernel_size=kernel_size
+        )
+
+    def remove_weight_norm(self):
+        self.encoder.remove_weight_norm()
+
+    @property
+    def input_types(self):
+        return {
+            "audio": NeuralType(('B', 'T_audio'), AudioSignal()),
+            "audio_len": NeuralType(tuple('B'), LengthsType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "encoded": NeuralType(('B', 'D', 'T_encoded'), EncodedRepresentation()),
+            "encoded_len": NeuralType(tuple('B'), LengthsType()),
+        }
+
+    @typecheck()
+    def forward(self, audio, audio_len):
+        encoded, encoded_len = self.pre_spec_processor(audio=audio, audio_len=audio_len)
+        encoded = self.pre_conv(inputs=encoded, input_len=encoded_len)
+        encoded = self.pre_res_block(inputs=encoded, input_len=encoded_len)
+
+        for stft_block in self.stft_blocks:
+            encoded, encoded_len = stft_block(
+                inputs=encoded, input_len=encoded_len, audio=audio, audio_len=audio_len
+            )
+
+        for down_sample_block in self.down_sample_blocks:
+            encoded, encoded_len = down_sample_block(inputs=encoded, input_len=encoded_len)
+
+        encoded = self.post_conv(inputs=encoded, input_len=encoded_len)
+
+        return encoded, encoded_len
+
+
+class VectorQuantizerIndexConverter(NeuralModule):
```

Comment:
Is there an eaiser way to do this? Can this be a function of the codec model or the quantizer part of the codec model itself?

```
        audio_file = "/home/audio.wav"
        codec_path = "/home/SpectralCodecFps43.nemo"
        audio, _ = librosa.load(audio_file, sr=sample_rate)
        audio_tensor = torch.tensor([audio])
        audio_len_tensor = torch.tensor([audio.shape[0]]).to(device)
        codec_model = AudioCodecModel.restore_from(codec_path)
        tokens, token_len = codec_model.encode(audio=audio_tensor, audio_len=audio_len_tensor)

        old_grouping = codec_model.get_grouping()
        codec_model.create_new_grouping(num_groups=6, num_levels_per_group=[5, 5, 5, 5])
```

Reply:
This is a temporary short-term implementation. If we decide that we want to keep this functionality long term, then I will look into a cleaner/faster way to do it when I implement the vectorized group FSQ class.

Reply:
This would work for on-the-fly extraction. It would not work if we precomputed the tokens and saved them on disk in a different structure.

I don't think VectorQuantizerIndexConverter() is a good solution, just a workaround to run the initial experiments. I will plan to add the generic conversion logic to the future vectorized FSQ implementation.